readhomer_atlas/library/importers.py

```diff
@@ -3,67 +3,65 @@
 
 from django.conf import settings
 
-from .models import Book, Line, Version
+from .models import Node
 
 
 LIBRARY_DATA_PATH = os.path.join(settings.PROJECT_ROOT, "data", "library")
 LIBRARY_METADATA_PATH = os.path.join(LIBRARY_DATA_PATH, "metadata.json")
 
 
-def _prepare_line_obj(version_obj, book_lookup, counters, line, line_idx):
+def _destructure_ref(reference):
+    """
+    NODE_HIERARCHY = ('A', 'B', 'C', 'D', 'E',)
+    assert _destructure_ref('1.2.3.a.5') == [
+        ('A', '1'),
+        ('B', '1.2'),
+        ('C', '1.2.3'),
+        ('D', '1.2.3.a'),
+        ('E', '1.2.3.a.5')
+    ]
+    """
+    components = reference.split(".")
+    return [
+        (kind, ".".join(components[: idx + 1]))
+        for idx, kind in enumerate(settings.NODE_HIERARCHY)
+    ]
+
+
+def _generate_branch(line, nodes, root_node):
     ref, tokens = line.strip().split(maxsplit=1)
     _, textpart_ref = ref.split(".", maxsplit=1)
-    book_ref, line_ref = textpart_ref.split(".", maxsplit=1)
-
-    book_obj = book_lookup.get(book_ref)
-    if book_obj is None:
-        book_obj, _ = Book.objects.get_or_create(
-            version=version_obj,
-            position=int(book_ref),
-            idx=counters["book_idx"],
-            ref=book_ref,
-            urn=Book.generate_urn(version_obj.urn, book_ref),
-        )
-        book_lookup[book_ref] = book_obj
-        counters["book_idx"] += 1
-    position = int(line_ref)
-    return Line(
-        text_content=tokens,
-        position=position,
-        idx=line_idx,
-        book=book_obj,
-        book_position=book_obj.position,
-        version=version_obj,
-        ref=textpart_ref,
-        urn=Line.generate_urn(version_obj.urn, textpart_ref),
-    )
+
+    refs = _destructure_ref(textpart_ref)
+    for idx, key in enumerate(refs):
+        parent = root_node if idx == 0 else nodes.get(refs[idx - 1])
+        node = nodes.get(key)
+        if node is None:
+            kind, ref = key
+            data = {"kind": kind, "urn": f"{root_node.urn}{ref}", "ref": ref}
+            if key == refs[-1]:
+                data.update({"text_content": tokens})
+            node = parent.add_child(**data)
```

Comment:
We might also look at using the `load_bulk` endpoint

https://django-treebeard.readthedocs.io/en/latest/api.html#treebeard.models.Node.load_bulk